trac_to_markdown.py

```diff
@@ -0,0 +1,89 @@
+"""Convert Trac's WikiFormat text to Markdown."""
+
+import re
+
+
+FLAVOURS = {
+    "markdown": {
+        "bold": r"**\1**",
+        "italic": r"*\1*",
+        "ul": r"*\2",
+        "ol": r"\1.\2",
+        "wiki_link": r"[%s/wiki/\2](\1)",
+        "ticket_link": r"[%s/ticket/\2](\1)",
+        "changeset_link": r"[%s/changeset/\2](\1)",
+    },
+    # mrkdwn is Slack's markup language, which is similar to Markdown.
+    "mrkdwn": {
+        "bold": r"*\1*",
+        "italic": r"_\1_",
+        "ul": r"\u2022\2",
+        "ol": r"\1\2",
+        "wiki_link": r"<%s/wiki/\1|\2>",
+        "ticket_link": r"<%s/ticket/\1|\2>",
+        "changeset_link": r"<%s/changeset/\1|\2>",
+    },
+}
+
+
+def convert(text, base="", flavour="markdown"):
+    """Convert the passed text from WikiFormatting to Markdown."""
+    transforms = FLAVOURS[flavour]
+
+    # Convert code blocks.
+    # XXX This doesn't handle nested blocks, but those should be
+    # XXX uncommon. We would probably need to properly parse the text
+    # XXX to handle that.
+    text = re.sub(r"^{{{\s(.*?)\s}}}$", r"```\n\1\n```", text, re.DOTALL)
+    # Also inline code:
+    text = re.sub(r"\{\{\{([^\n]+?)\}\}\}", r"`\1`", text)
+
+    # Convert bold and italic text.
+    # Note that it's important to do bold before italic, to avoid having
+    # to distinguish them.
+    text = re.sub(r"'''(.+)'''", transforms["bold"], text)
+    text = re.sub(r"''(.+)''", transforms["italic"], text)
+    # Backticks ("monospaced" in Trac terms) and quotes
+    # ("Discussion citations") are the same so do not need conversion.
+
+    # TODO: Definition lists.
+    # TODO: Blockquotes (maybe these are identical?).
+    # TODO: Tables (would need MD extension).
+    # TODO: Images.
+    # TODO: Line breaks.
+    # TODO: Macros (don't know what could be done with these, though).
+
+    # Convert headings.
+    # Again, note that the order is important to keep this simple.
+    # Slack will not handle these specially, but they look nicer in
+    # markdown, so keep the same conversion.
+    text = re.sub(r"====\s+(.+?)\s+====", r"#### \1", text)
+    text = re.sub(r"===\s+(.+?)\s+===", r"### \1", text)
+    text = re.sub(r"==\s+(.+?)\s+==", r"## \1", text)
+    text = re.sub(r"=\s+(.+?)\s+=", r"# \1", text)
+
+    # Convert lists.
+    text = re.sub(r"(^\s\*)(\s)", transforms["ul"], text)
+    text = re.sub(r"(^\s\-)(\s)", transforms["ul"], text)
+    text = re.sub(r"^\s(\d+)\.(\s)", transforms["ol"], text)
+
```

Comment:
Also these 3 seems to need `flags=re.M` since the `^` will only match at the start of string otherwise.